lib/ts/recipe/totp/recipeImplementation.ts

```diff
@@ -0,0 +1,176 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { RecipeInterface } from "./";
+import { Querier } from "../../querier";
+import NormalisedURLPath from "../../normalisedURLPath";
+import { TypeNormalisedInput } from "./types";
+import { UserContext } from "../../types";
+import { getUser } from "../..";
+
+export default function getRecipeInterface(querier: Querier, config: TypeNormalisedInput): RecipeInterface {
+    return {
+        getUserIdentifierInfoForUserId: async function (this: RecipeInterface, { userId, userContext }) {
+            let user = await getUser(userId, userContext);
+
+            if (user === undefined) {
+                return {
+                    status: "UNKNOWN_USER_ID_ERROR",
+                };
+            }
+
+            const primaryLoginMethod = user.loginMethods.find(
+                (method) => method.recipeUserId.getAsString() === user!.id
+            );
+
+            if (primaryLoginMethod !== undefined) {
+                if (primaryLoginMethod.email !== undefined) {
+                    return {
+                        info: primaryLoginMethod.email,
+                        status: "OK",
+                    };
+                } else if (primaryLoginMethod.phoneNumber !== undefined) {
+                    return {
+                        info: primaryLoginMethod.phoneNumber,
+                        status: "OK",
+                    };
+                }
+            }
+
+            if (user.emails.length > 0) {
+                return { info: user.emails[0], status: "OK" };
+            } else if (user.phoneNumbers.length > 0) {
+                return { info: user.phoneNumbers[0], status: "OK" };
+            }
+
+            return {
+                status: "USER_IDENTIFIER_INFO_DOES_NOT_EXIST_ERROR",
+            };
+        },
+
+        createDevice: async function (
+            this: RecipeInterface,
+            input: {
+                userId: string;
+                deviceName?: string;
+                skew?: number;
+                period?: number;
+                userIdentifierInfo?: string;
+                userContext: UserContext;
+            }
+        ) {
+            if (input.userIdentifierInfo === undefined) {
+                const emailOrPhoneInfo = await this.getUserIdentifierInfoForUserId({
+                    userId: input.userId,
+                    userContext: input.userContext,
+                });
+                if (emailOrPhoneInfo.status === "OK") {
+                    input.userIdentifierInfo = emailOrPhoneInfo.info;
+                } else {
+                    // Ignore since UserIdentifierInfo is optional
```

Comment:
UNKONWN_USER_ID should be propagated as a status code. The API calling this should handle this by throwing an unauthorised error, and the index function calling this should return back the status code to the dev.

Reply:
Done